xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java

```diff
@@ -747,4 +973,121 @@ private void notifyWatcher(ResourceWatcher<T> watcher, T update) {
+
+      ControlPlaneClient cpcClosed = serverCpClientMap.get(serverInfo);
+      if (cpcClosed == null) {
+        return;
+      }
+
+      if (!inRetry) {
+        cleanUpResourceTimers(cpcClosed);
+      }
+      if (cpcClosed.isResponseRecieved()) {
+        return; // Not considered an error
+      }
+
+      Status error = status.isOk() ? Status.UNAVAILABLE.withDescription(CLOSED_BY_SERVER) : status;
+      boolean checkForFallback = !status.isOk() && !inRetry;
+
+      List<String> authoritiesForClosedCpc = getActiveAuthorities(cpcClosed);
+      for (Map<String, ResourceSubscriber<? extends ResourceUpdate>> subscriberMap :
+          resourceSubscribers.values()) {
+        for (ResourceSubscriber<? extends ResourceUpdate> subscriber : subscriberMap.values()) {
+          if (subscriber.hasResult() || !authoritiesForClosedCpc.contains(subscriber.authority)) {
+            continue;
+          }
+
+          // try to fallback to lower priority control plane client
+          if (checkForFallback
+              && doFallbackForAuthority(cpcClosed, subscriber.serverInfos, subscriber.authority)) {
+            authoritiesForClosedCpc.remove(subscriber.authority);
+            if (authoritiesForClosedCpc.isEmpty()) {
+              return; // optimization: no need to continue once all authorities have done fallback
+            }
+            continue; // since we did fallback, don't consider it an error
+          }
+
+          subscriber.onError(error, null);
+        }
+      }
+    }
+
+    @Override
+    public void handleStreamRestarted(ServerInfo serverInfo) {
+      syncContext.throwIfNotInThisSynchronizationContext();
+
+      if (isShutDown()) {
+        return;
+      }
+
+      ControlPlaneClient controlPlaneClient = serverCpClientMap.get(serverInfo);
+      if (controlPlaneClient == null) {
+        return;
+      }
+
+      for (String authority : getAuthoritiesForServerInfo(serverInfo)) {
+        if (hasSubscribers(authority)) {
+          internalHandleStreamReady(controlPlaneClient, authority);
+        }
+      }
+    }
+
+    private void internalHandleStreamReady(ControlPlaneClient readyCpc, String authority) {
+      ControlPlaneClient activeCpClient = getActiveCpc(authority);
+      if (activeCpClient == null) {
+        activeCpClients.computeIfAbsent(authority, k -> new ArrayList<>()).add(readyCpc);
+        restartMatchingSubscriberTimers(authority);
+        readyCpc.sendDiscoveryRequests();
+        return; // Since nothing was active can ignore fallback
+      }
+
+      if (activeCpClient.isReady()
+          && !activeCpClients.get(authority).contains(readyCpc)) {
+        logger.log(XdsLogLevel.INFO, "Ignoring stream restart for lower priority server {0}",
+            readyCpc.getServerInfo().target());
+        return;
+      }
+
+      if (activeCpClient != readyCpc) {
+        readyCpc.markStartingUp();
+      }
+
+      restartMatchingSubscriberTimers(authority);
+      readyCpc.sendDiscoveryRequests();
```

Comment:
This isn't limited to the current `authority`, so won't this send discovery multiple times?